playground/backend/internal/code_processing/code_processing.go

```diff
@@ -250,6 +268,54 @@ func cancelCheck(ctx context.Context, pipelineId uuid.UUID, cancelChannel chan b
 	}
 }
 
+// readLogFile reads logs from the log file and keeps it to the cache.
+// If context is done it means that code processing was finished (successfully/with error/timeout). Write last logs to the cache.
+// If <-stopReadLogsChannel it means that code processing was finished (canceled/timeout)
+// 	and it waits when this method stop working to change status to the pb.Status_STATUS_FINISHED. Write last logs
+//	to the cache and set value to the finishReadLogChannel channel to unblock code processing.
+// In other case each 500 milliseconds write to cache logs of the code processing.
+func readLogFile(ctx context.Context, cacheService cache.Cache, logFilePath string, pipelineId uuid.UUID, stopReadLogsChannel, finishReadLogChannel chan bool) {
+	ticker := time.NewTicker(500 * time.Millisecond)
+	for {
+		select {
+		// in case of timeout or cancel
+		case <-ctx.Done():
+			finishReadLogFile(ctx, ticker, cacheService, logFilePath, pipelineId)
+			return
+		// in case of pipeline finish successfully or has error on the run step
+		case <-stopReadLogsChannel:
+			finishReadLogFile(ctx, ticker, cacheService, logFilePath, pipelineId)
+			finishReadLogChannel <- true
+			return
+		case <-ticker.C:
+			writeLogsToCache(ctx, cacheService, logFilePath, pipelineId)
+		}
+	}
+}
+
+// finishReadLogFile is used to read logs file for the last time
+func finishReadLogFile(ctx context.Context, ticker *time.Ticker, cacheService cache.Cache, logFilePath string, pipelineId uuid.UUID) {
+	ticker.Stop()
+	writeLogsToCache(ctx, cacheService, logFilePath, pipelineId)
+}
+
+// writeLogsToCache write all logs from the log file to the cache.
+// If log file doesn't exist, return.
+//	Reading logs works as a parallel with code processing so when program tries to read file
+//	it could be that the file doesn't exist yet.
+// If log file exists, read all from the log file and keep it to the cache using cache.Logs subKey.
+// If some error occurs, log the error and return.
+func writeLogsToCache(ctx context.Context, cacheService cache.Cache, logFilePath string, pipelineId uuid.UUID) {
+	if _, err := os.Stat(logFilePath); os.IsNotExist(err) {
+		return
+	}
+	logs, err := os.ReadFile(logFilePath)
```

Comment:
I remember I had a question about os.Stat(logFilePath);

Reply:
If file doesn't exist it is not the error case. In some cases, this method could start to read the file before this file is created by code processing (this method works synchronously with code processing). So if the file doesn't exist need to try after some pause.